File: brain/memory/vector_store.py
"""
Vector Store for Memory Storage (Chunk 2.4)
Uses RedisVL for vector storage and sentence-transformers for embeddings.
"""

import time
import json
from pathlib import Path
from typing import List, Dict, Any, Optional

import redis
from sentence_transformers import SentenceTransformer
import numpy as np


class VectorStore:
    """Vector store for memory storage using RedisVL"""

    def __init__(self, redis_url: str = "redis://localhost:6379"):
        self.redis_url = redis_url
        self.client: Optional[redis.Redis] = None
        self.model: Optional[SentenceTransformer] = None
        self.index_name = "memory_index"
        self.prefix = "memory:"
        self.embedding_dim = 384  # all-MiniLM-L6-v2 dimension

    def connect(self):
        """Initialize Redis connection and sentence transformer model"""
        self.client = redis.from_url(
            self.redis_url,
            decode_responses=False  # Binary mode for vectors
        )

        # Load sentence-transformers model (CPU)
        print("[VectorStore] Loading sentence-transformers model...")
        self.model = SentenceTransformer('all-MiniLM-L6-v2')
        print("[VectorStore] Model loaded")

    def create_index(self):
        """
        Create RediSearch index for memory storage.
        Idempotent - skips if index already exists.
        """
        try:
            # Check if index exists
            try:
                self.client.ft(self.index_name).info()
                print(f"[VectorStore] Index '{self.index_name}' already exists")
                return
            except redis.exceptions.ResponseError:
                # Index doesn't exist, create it
                pass

            # Create index using FT.CREATE command
            # Schema: message (TEXT), embedding (VECTOR), perplexity (NUMERIC),
            #         surprise_score (NUMERIC), timestamp (NUMERIC),
            #         session_id (TAG), metadata (TEXT)
            self.client.execute_command(
                "FT.CREATE", self.index_name,
                "ON", "HASH",
                "PREFIX", "1", self.prefix,
                "SCHEMA",
                "message", "TEXT",
                "embedding", "VECTOR", "HNSW", "6",
                "TYPE", "FLOAT32",
                "DIM", str(self.embedding_dim),
                "DISTANCE_METRIC", "COSINE",
                "perplexity", "NUMERIC", "SORTABLE",
                "surprise_score", "NUMERIC", "SORTABLE",
                "timestamp", "NUMERIC", "SORTABLE",
                "session_id", "TAG",
                "metadata", "TEXT"
            )

            print(f"[VectorStore] Created index '{self.index_name}'")

        except Exception as e:
            print(f"[VectorStore] Error creating index: {e}")
            raise

    def close(self):
        """Close Redis connection"""
        if self.client:
            self.client.close()

    def generate_embedding(self, text: str) -> np.ndarray:
        """
        Generate embedding vector for text using sentence-transformers.

        Args:
            text: Input text to embed

        Returns:
            Embedding vector (384 dimensions)
        """
        if not self.model:
            raise RuntimeError("Model not loaded. Call connect() first.")

        embedding = self.model.encode(text, convert_to_numpy=True)
        return embedding

    def store_memory(
        self,
        message: str,
        perplexity: float,
        surprise_score: Optional[float] = None,
        session_id: str = "default",
        metadata: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Store a memory with vector embedding.

        Args:
            message: The text message to store
            perplexity: Perplexity score for the message
            surprise_score: Optional surprise score (perplexity + vector distance)
            session_id: Session identifier
            metadata: Optional metadata dict

        Returns:
            Redis key of stored memory
        """
        # Generate embedding
        embedding = self.generate_embedding(message)

        # Create memory key
        timestamp = time.time()
        memory_key = f"{self.prefix}{int(timestamp * 1000)}"

        # Prepare memory data
        memory_data = {
            "message": message,
            "embedding": embedding.astype(np.float32).tobytes(),
            "perplexity": perplexity,
            "surprise_score": surprise_score if surprise_score else 0.0,
            "timestamp": timestamp,
            "session_id": session_id,
            "metadata": json.dumps(metadata) if metadata else "{}"
        }

        # Store in Redis
        self.client.hset(memory_key, mapping=memory_data)

        print(f"[VectorStore] Stored memory: {memory_key}")
        return memory_key

    def search_similar(
        self,
        query_text: str,
        top_k: int = 5,
        session_id: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        Search for similar memories using vector similarity.

        Args:
            query_text: Text to search for
            top_k: Number of results to return
            session_id: Optional session filter

        Returns:
            List of similar memories with scores
        """
        # Generate query embedding
        query_embedding = self.generate_embedding(query_text)
        query_bytes = query_embedding.astype(np.float32).tobytes()

        # Build KNN query
        if session_id:
            # Escape special characters in tag value
            escaped_session = session_id.replace("_", "\\_")
            base_query = f"(@session_id:{{{escaped_session}}})=>[KNN {top_k} @embedding $vec AS score]"
        else:
            base_query = f"*=>[KNN {top_k} @embedding $vec AS score]"

        # Execute search using FT.SEARCH
        try:
            results = self.client.execute_command(
                "FT.SEARCH", self.index_name,
                base_query,
                "PARAMS", "2", "vec", query_bytes,
                "RETURN", "6", "message", "perplexity",
                "surprise_score", "timestamp", "session_id", "score",
                "SORTBY", "score",
                "DIALECT", "2",
                "LIMIT", "0", str(top_k)
            )
        except Exception as e:
            print(f"[VectorStore] Search error: {e}")
            return []

        # Parse results
        # Format: [count, key1, fields1, key2, fields2, ...]
        memories = []
        if not results or results[0] == 0:
            return memories

        # Skip the count, iterate over key-fields pairs
        for i in range(1, len(results), 2):
            key = results[i]
            fields = results[i + 1]

            # Decode key if bytes
            if isinstance(key, bytes):
                key = key.decode()

            # Convert field list to dict
            field_dict = {}
            for j in range(0, len(fields), 2):
                field_name = fields[j]
                field_value = fields[j + 1]

                # Decode bytes to strings
                if isinstance(field_name, bytes):
                    field_name = field_name.decode()
                if isinstance(field_value, bytes):
                    field_value = field_value.decode()

                field_dict[field_name] = field_value

            memories.append({
                "id": key,
                "message": field_dict.get("message", ""),
                "perplexity": float(field_dict.get("perplexity", 0)),
                "surprise_score": float(field_dict.get("surprise_score", 0)),
                "timestamp": float(field_dict.get("timestamp", 0)),
                "session_id": field_dict.get("session_id", ""),
                "similarity_score": float(field_dict.get("score", 1.0))
            })

        return memories


# Global instance
vector_store = VectorStore()

File: archive-ai-rs/crates/brain/src/services/vector_store.rs
use redis::{AsyncCommands, cmd};
use serde_json::json;
use crate::services::redis_client::RedisService;
use shared::AppConfig;
use anyhow::Result;
use tracing::{info, error, debug};

#[derive(Clone)]
pub struct VectorStore {
    redis: RedisService,
    config: AppConfig,
    client: reqwest::Client,
    index_name: String,
    prefix: String,
}

impl VectorStore {
    pub fn new(redis: RedisService, config: AppConfig) -> Self {
        Self {
            redis,
            config,
            client: reqwest::Client::new(),
            index_name: "memory_index".to_string(),
            prefix: "memory:".to_string(),
        }
    }

    pub async fn create_index(&self) -> Result<()> {
        let mut conn = self.redis.get_connection().await?;
        
        // Check if index exists
        let info: Result<redis::Value, _> = cmd("FT.INFO")
            .arg(&self.index_name)
            .query_async(&mut conn)
            .await;

        if info.is_ok() {
            info!("Index '{}' already exists", self.index_name);
            return Ok(());
        }

        info!("Creating index '{}'...", self.index_name);
        
        // FT.CREATE memory_index ON HASH PREFIX 1 memory: SCHEMA ...
        let _: () = cmd("FT.CREATE")
            .arg(&self.index_name)
            .arg("ON").arg("HASH")
            .arg("PREFIX").arg("1").arg(&self.prefix)
            .arg("SCHEMA")
            .arg("message").arg("TEXT")
            .arg("embedding").arg("VECTOR").arg("HNSW").arg("6")
                .arg("TYPE").arg("FLOAT32")
                .arg("DIM").arg("384")
                .arg("DISTANCE_METRIC").arg("COSINE")
            .arg("perplexity").arg("NUMERIC").arg("SORTABLE")
            .arg("surprise_score").arg("NUMERIC").arg("SORTABLE")
            .arg("timestamp").arg("NUMERIC").arg("SORTABLE")
            .arg("session_id").arg("TAG")
            .arg("metadata").arg("TEXT")
            .query_async(&mut conn)
            .await?;

        Ok(())
    }

    pub async fn get_embedding(&self, text: &str) -> Result<Vec<f32>> {
        let url = format!("{}/v1/embeddings", self.config.vorpal_url);
        let payload = json!({
            "input": text,
            "model": self.config.vorpal_model
        });

        let res = self.client.post(&url)
            .json(&payload)
            .send()
            .await?;

        if !res.status().is_success() {
            return Err(anyhow::anyhow!("Embedding service error: {}", res.status()));
        }

        let body: serde_json::Value = res.json().await?;
        
        let embedding = body["data"][0]["embedding"]
            .as_array()
            .ok_or(anyhow::anyhow!("Invalid embedding response format"))?
            .iter()
            .map(|v| v.as_f64().unwrap_or(0.0) as f32)
            .collect();

        Ok(embedding)
    }

    pub async fn search_similar(&self, query_vec: &[f32], limit: usize) -> Result<Vec<serde_json::Value>> {
        let mut conn = self.redis.get_connection().await?;
        
        // Convert Vec<f32> to bytes
        let vec_bytes: Vec<u8> = query_vec.iter()
            .flat_map(|&f| f.to_le_bytes().to_vec())
            .collect();

        let query = format!("*=>[KNN {} @embedding $vec AS score]", limit);

        let results: redis::Value = cmd("FT.SEARCH")
            .arg(&self.index_name)
            .arg(query)
            .arg("PARAMS").arg("2").arg("vec").arg(vec_bytes)
            .arg("RETURN").arg("6").arg("message").arg("perplexity").arg("surprise_score").arg("timestamp").arg("session_id").arg("score")
            .arg("SORTBY").arg("score")
            .arg("DIALECT").arg("2")
            .arg("LIMIT").arg("0").arg(limit.to_string())
            .query_async(&mut conn)
            .await?;

        // Parse RediSearch results (Format: [count, key1, [fields1], key2, [fields2]...])
        // This parsing is tedious in Rust but necessary.
        // For brevity, I'll implement a basic parser.
        
        match results {
            redis::Value::Bulk(items) => {
                if items.is_empty() { return Ok(vec![]); }
                // let count = match &items[0] { redis::Value::Int(c) => *c, _ => 0 };
                let mut memories = Vec::new();
                
                for i in (1..items.len()).step_by(2) {
                    // let key = &items[i];
                    let fields = &items[i+1];
                    
                    if let redis::Value::Bulk(f_items) = fields {
                        let mut map = serde_json::Map::new();
                        for j in (0..f_items.len()).step_by(2) {
                            if let (redis::Value::Data(name), redis::Value::Data(val)) = (&f_items[j], &f_items[j+1]) {
                                let name_str = String::from_utf8_lossy(name).to_string();
                                let val_str = String::from_utf8_lossy(val).to_string();
                                map.insert(name_str, serde_json::Value::String(val_str));
                            }
                        }
                        memories.push(serde_json::Value::Object(map));
                    }
                }
                Ok(memories)
            },
            _ => Ok(vec![])
        }
    }
    
    pub async fn store_memory(
        &self, 
        message: &str, 
        perplexity: f64, 
        surprise_score: f64,
        session_id: &str,
        metadata: serde_json::Value
    ) -> Result<String> {
        let mut conn = self.redis.get_connection().await?;
        
        let embedding = self.get_embedding(message).await?;
        let vec_bytes: Vec<u8> = embedding.iter()
            .flat_map(|&f| f.to_le_bytes().to_vec())
            .collect();

        let timestamp = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)?
            .as_secs_f64();
            
        let key = format!("{}{}", self.prefix, (timestamp * 1000.0) as u64);

        let _: () = conn.hset_multiple(&key, &[
            ("message", message),
            ("perplexity", &perplexity.to_string()),
            ("surprise_score", &surprise_score.to_string()),
            ("timestamp", &timestamp.to_string()),
            ("session_id", session_id),
            ("metadata", &metadata.to_string()),
        ]).await?;
        
        // Set binary embedding separately to avoid string conversion issues in hset_multiple
        let _: () = conn.hset(&key, "embedding", vec_bytes).await?;

        debug!("Stored memory: {}", key);
        Ok(key)
    }
}